src/building/bootstrapping.md

```diff
@@ -122,50 +124,109 @@ contribution [here][bootstrap-build].
 
 ## Understanding stages of bootstrap
 
-This is a detailed look into the separate bootstrap stages. When running
-`x.py` you will see output such as:
-
-```txt
-Building stage0 std artifacts
-Copying stage0 std from stage0
-Building stage0 compiler artifacts
-Copying stage0 rustc from stage0
-Building LLVM for x86_64-apple-darwin
-Building stage0 codegen artifacts
-Assembling stage1 compiler
-Building stage1 std artifacts
-Copying stage1 std from stage1
-Building stage1 compiler artifacts
-Copying stage1 rustc from stage1
-Building stage1 codegen artifacts
-Assembling stage2 compiler
-Uplifting stage1 std
-Copying stage2 std from stage1
-Generating unstable book md files
-Building stage0 tool unstable-book-gen
-Building stage0 tool rustbook
-Documenting standalone
-Building rustdoc for stage2
-Documenting book redirect pages
-Documenting stage2 std
-Building rustdoc for stage1
-Documenting stage2 whitelisted compiler
-Documenting stage2 compiler
-Documenting stage2 rustdoc
-Documenting error index
-Uplifting stage1 rustc
-Copying stage2 rustc from stage1
-Building stage2 tool error_index_generator
-```
+### Overview
+
+This is a detailed look into the separate bootstrap stages.
+
+The convention `x.py` uses is that:
+- A `--stage N` flag means to run the stage N compiler (`stageN/rustc`).
+- A "stage N artifact" is an artifact that is _produced_ by the stage N compiler.
+- The "stage (N+1) compiler" is assembled from "stage N artifacts". This
+  process is called _uplifting_.
+
+#### Examples
+
+- `x.py build --stage 0` means to build with the beta `rustc`.
+- `x.py doc --stage 0` means to document using the beta `rustdoc`.
+- `x.py test --stage 0 library/std` means to run tests on the standard library
+  without building `rustc` from source ('build with stage 0, then test the artifacts').
+- `x.py test --stage 0 src/test/ui` is not meaningful: it runs tests on the
+  _beta_ compiler and doesn't build `rustc` from source. Use `test src/test/ui`
+  instead, which builds from source.
+- `x.py test --stage 0 compiler/rustc` builds the compiler but runs no tests:
+  it's running `cargo test -p rustc`, but cargo doesn't understand Rust's
+  tests. You shouldn't need to use this, use `test` instead (without arguments).
+
+### Building vs. Running
+
+Note that this implies the stage N compiler is _not_ the same as the compiler
+built by `build --stage N compiler/rustc` -- that's 'stage N artifacts'
+('the compiler built by stage N').
```

Comment:
Suggestion:
```suggestion
Note that this implies that `build --stage N compiler/rustc` does _not_ build
the stage N compiler: `--stage N` means to build 'stage N' artifacts, and
the compiler built with stage N artifacts is actually the stage (N+1) compiler.
```

Reply:
This is incorrect and IMO more confusing than the original. 'built with stage N artifacts' doesn't really make sense, but even if you meant 'the stage N+1 compiler' it doesn't make sense to say stage N+1 builds itself.

Maybe something like this?
```suggestion
Note that this implies that the stage N compiler is different from `build --stage N compiler/rustc`: `--stage N` means to build _using_ the stage N compiler, while `compiler/rustc` is the 'stage N artifacts', which will eventually be uplifted to stage (N+1).
```

Reply:
What about this? It's shorter and simpler. I'm not sure if I cut out some information you want to share, but if that's true maybe have the first sentence be this and then further explain?

```suggestion
Note that `build --stage N compiler/rustc` **does not** build the stage N compiler:
instead it builds the stage _N+1_ compiler _using_ the stage N compiler.
```

I wasn't sure what "this" was referring in "Note that this implies", so I cut that part out.

Reply:
Yeah, I like that a lot better. Thanks!